### ISSUE TITLE
Support best_of_n method based on reward model  on Scaffolding

### ISSUE THREAD

**Issue by WeiHaocheng** [NVIDIA/TensorRT-LLM#3160]
Support best_of_n method based on reward model  on Scaffolding

**Comment by linuxlonelyeagle** [comment]
It looks interesting.I'd be happy to finish the work.

**Comment by WeiHaocheng** [comment]
Great~ 

- For the use of reward models， you can take a look on [here](https://github.com/NVIDIA/TensorRT-LLM/blob/36ac5e78eddb889b1332ca1f15eecfa7972bf049/tests/unittest/llmapi/test_llm.py#L2052).


- Due to the need to run two models(generation model and reward model), it requires the use of the openai api worker on scaffolding. It depends on this [PR](https://github.com/NVIDIA/TensorRT-LLM/pull/3166).

**Comment by WeiHaocheng** [comment]
How to start PRM server:

[To start server]
7B PRM
trtllm-serve /scratch/llm-models/Qwen2.5-Math-PRM-7B/ --backend pytorch --host 0.0.0.0 --port 30000 --extra_llm_api_options ./extra-llm-api-config.yml
72B ORM
trtllm-serve /scratch/llm-models/Qwen2.5-Math-RM-72B/ --backend pytorch --tp_size 2  --host 0.0.0.0 --port 30000 --extra_llm_api_options ./extra-llm-api-config.yml


In ./extra-llm-api-config.yml , it needs to set attn backend:
pytorch_backend_config:
    attn_backend: "VANILLA"

# example query (only /v1/completions supports return logits)
curl http://localhost:30000/v1/completions \ost:30000/v1/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen2.5-Math-PRM-7B",
    "prompt": "A B C",
    "return_context_logits": true
  }'
# example output
{
    "id": "cmpl-878ee25b66d04e938dd8e95ce826665b",
    "object": "text_completion",
    "created": 1742254830,
    "model": "Qwen2.5-Math-PRM-7B",
    "choices": [
        {
            "index": 0,
            "text": "",
            "logprobs": null,
            "context_logits": [    <-------------scoring logits are here
                [-0.2890625, [0.0927734375](tel:00927734375)],
                [-0.71875, -0.96875],
                [-0.400390625, -0.058349609375]
            ],
            "finish_reason": "length",
            "stop_reason": null,
            "disaggregated_params": null
        }
    ],
    "usage": {
        "prompt_tokens": 3,
        "total_tokens": 3,
        "completion_tokens": 0
    }
}

**Comment by narutolhy** [comment]
Good, I want to support this feature, please assign to me. Thanks

**Comment by WeiHaocheng** [comment]
About how to get reward value with Qwen2.5-Math-PRM-7B, we can take a look into [here](https://github.com/RyanLiu112/compute-optimal-tts/blob/0ee2578af1f8d6cac445c9c4c72780528bb94556/src/reason/inference/infer_fns.py#L118).

**Comment by WeiHaocheng** [comment]
As we want to support best of n in release 0.19, so we assign zhenhuan @dc3671 to support an basic case for best of N.

**Comment by WeiHaocheng** [comment]
Some points from my thought:

- Erin has support and verify Qwen2.5-Math-PRM-7B, other reward models have not been validated yet, so we may begin with this model.

- About how to calculate the reward score from output logits of Qwen2.5-Math-PRM-7B, there is a case from compute-optimal-tts:

```
rewards = []
for conversation in conversations:
    
    input_ids = tokenizer.apply_chat_template(conversation, return_tensors="pt").to(device)
    indices = (input_ids == special_tag_id)
    logits = model(input_ids)[0]
    scores = logits.softmax(dim=-1)[0]
    probabilities = scores * indices[0].unsqueeze(-1)
    step_scores = probabilities[probabilities != 0].view(-1, 2)[:, 1]
    rewards.append(copy.deepcopy(step_scores))
```

See more detail on [here](https://github.com/RyanLiu112/compute-optimal-tts/blob/0ee2578af1f8d6cac445c9c4c72780528bb94556/src/reason/inference/infer_fns.py#L118)~

- We may need to use TRTOpenaiWorker to run generation model and reward model in the same time. And we need set --kv_cache_free_gpu_memory_fraction as 0.5 or less, otherwise the first service to start will occupy all gpu memory.

**Comment by therealnaveenkamal** [comment]
@WeiHaocheng is this done? I didn't see any comments/PR. Can I work on this?
